### PR TITLE
RSDK-3950 - call into cgo for getPosition

### PR DESCRIPTION
**Pull request by kim-mishra**
[RSDK-3950](https://viam.atlassian.net/browse/RSDK-3950)

- If the modularizationV2Enabled feature flag is true, call into the new cartofacade API for getPosition
- Write unit tests to ensure new conversion returns the same values as the old conversions would have

[RSDK-3950]: https://viam.atlassian.net/browse/RSDK-3950?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: viam-cartographer.go [hunk 1]
```diff
@@ -509,6 +514,24 @@ func (cartoSvc *cartographerService) GetPosition(ctx context.Context) (spatialma
 	return vcUtils.CheckQuaternionFromClientAlgo(pose, componentReference, returnedExt)
 }
 
+func (cartoSvc *cartographerService) getPositionModularizationV2(ctx context.Context) (spatialmath.Pose, string, error) {
+	pos, err := cartoSvc.cartofacade.GetPosition(ctx, cartoSvc.cartoFacadeTimeout)
+	if err != nil {
+		return spatialmath.NewPose(r3.Vector{}, spatialmath.NewOrientationVector()), "", err
```
Comment (nicksanford) [inline viam-cartographer.go:520]: why not:
```
return spatialmath.NewZeroPose(), "", err
```

Comment (kim-mishra) [reply viam-cartographer.go]: I must've missed that function - udpdated!

File: viam-cartographer.go [hunk 2]
```diff
@@ -509,6 +514,24 @@ func (cartoSvc *cartographerService) GetPosition(ctx context.Context) (spatialma
 	return vcUtils.CheckQuaternionFromClientAlgo(pose, componentReference, returnedExt)
 }
 
+func (cartoSvc *cartographerService) getPositionModularizationV2(ctx context.Context) (spatialmath.Pose, string, error) {
+	pos, err := cartoSvc.cartofacade.GetPosition(ctx, cartoSvc.cartoFacadeTimeout)
+	if err != nil {
+		return spatialmath.NewZeroPose(), "", err
```
Comment (nicksanford) [inline viam-cartographer.go:520]: actually, can this just be `return nil, "", err`?

File: viam-cartographer_internal_test.go [hunk 3]
```diff
@@ -243,6 +244,75 @@ func TestGetPositionEndpoint(t *testing.T) {
...
+			pose, _, err := svc.GetPosition(context.Background())
+			test.That(t, err, test.ShouldBeNil)
+			expectedPose := spatialmath.NewPose(
+				r3.Vector{X: inputPose.X, Y: inputPose.Y, Z: inputPose.Z},
+				makeQuaternionFromGenericMap(inputQuat),
+			)
+			test.That(t, pose, test.ShouldResemble, expectedPose)
+		})
+
+		t.Run("non origin pose success", func(t *testing.T) {
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:286]: can we test the error case please?

File: viam-cartographer_internal_test.go [hunk 4]
```diff
@@ -243,6 +244,75 @@ func TestGetPositionEndpoint(t *testing.T) {
 	})
 }
 
+//nolint:dupl
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:247]: is this still necessary if we remove 257?

Comment (kim-mishra) [reply viam-cartographer_internal_test.go:247]: With 257 removed, I still get: ` 284-310 lines are duplicate of `viam-cartographer_internal_test.go:256-282``

Comment (nicksanford) [reply viam-cartographer_internal_test.go:247]: can you please refactor it to not hit this lint rule by making a helper function?

File: viam-cartographer.go [hunk 5]
```diff
@@ -509,6 +514,24 @@ func (cartoSvc *cartographerService) GetPosition(ctx context.Context) (spatialma
 	return vcUtils.CheckQuaternionFromClientAlgo(pose, componentReference, returnedExt)
 }
 
+func (cartoSvc *cartographerService) getPositionModularizationV2(ctx context.Context) (spatialmath.Pose, string, error) {
+	pos, err := cartoSvc.cartofacade.GetPosition(ctx, cartoSvc.cartoFacadeTimeout)
+	if err != nil {
+		return nil, "", err
```
Comment (nicksanford) [inline viam-cartographer.go:520]: can we please test this error case?

File: viam-cartographer_internal_test.go [hunk 6]
```diff
@@ -243,6 +244,73 @@ func TestGetPositionEndpoint(t *testing.T) {
...
+					Imag: 0.0,
+					Jmag: 0.0,
+					Kmag: 0.0,
+				},
+				nil
+		}
+
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		pose, _, err := svc.GetPosition(context.Background())
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:275]: why are we not testing the component reference?

File: viam-cartographer_internal_test.go [hunk 7]
```diff
@@ -243,6 +244,73 @@ func TestGetPositionEndpoint(t *testing.T) {
...
+					Imag: 1.0,
+					Jmag: 0.0,
+					Kmag: 0.0,
+				},
+				nil
+		}
+
+		inputPose = commonv1.Pose{X: 5, Y: 5, Z: 5, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 1.0, "jmag": 0.0, "kmag": 0.0}
+		pose, _, err := svc.GetPosition(context.Background())
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:303]: why are we not testing the component reference?

File: viam-cartographer_internal_test.go [hunk 8]
```diff
@@ -243,6 +244,73 @@ func TestGetPositionEndpoint(t *testing.T) {
 	})
 }
 
+//nolint:dupl
+func TestGetPositionModularizationV2Endpoint(t *testing.T) {
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:247]: We need a test for `"empty component reference success"`

File: viam-cartographer_internal_test.go (hunk 6 above)
Comment (kim-mishra) [reply viam-cartographer_internal_test.go:275]: I will add that back in, I got confused about using this after some of the changes from component reference to sensors[0], but this will still be needed after full mod in the case of multiple sensors, right?

File: viam-cartographer_internal_test.go [hunk 9]
```diff
@@ -243,6 +244,88 @@ func TestGetPositionEndpoint(t *testing.T) {
...
+		mockCartoFacade.GetPositionFunc = func(
+			ctx context.Context,
+			timeout time.Duration,
+		) (cartofacade.GetPosition, error) {
+			return cartofacade.GetPosition{}, errors.New("testError")
+		}
+
+		inputPose = commonv1.Pose{X: 0, Y: 0, Z: 0, OX: 0, OY: 0, OZ: 1, Theta: 0}
+		inputQuat = map[string]interface{}{"real": 1.0, "imag": 0.0, "jmag": 0.0, "kmag": 0.0}
+		pose, _, err := svc.GetPosition(context.Background())
```
Comment (nicksanford) [inline viam-cartographer_internal_test.go:322]: why are we not testing the component reference?

File: viam-cartographer_internal_test.go (hunk 6 above)
Comment (nicksanford) [reply viam-cartographer_internal_test.go:275]: yes correct

File: viam-cartographer_internal_test.go (hunk 7 above)
Comment (kim-mishra) [reply viam-cartographer_internal_test.go]: fixed

File: viam-cartographer_internal_test.go (hunk 9 above)
Comment (kim-mishra) [reply viam-cartographer_internal_test.go]: fixed

**Review by nicksanford — APPROVED** [review]
